From: Simon McVittie <smcv@debian.org>
Date: Sun, 26 Jul 2026 11:48:21 +0100
Subject: tests: Avoid a warning with newer GLib

Signed-off-by: Simon McVittie <smcv@debian.org>
Origin: upstream, 0.1.8
---
 tests/test-proxy.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/tests/test-proxy.c b/tests/test-proxy.c
index df1f00b..ae106d9 100644
--- a/tests/test-proxy.c
+++ b/tests/test-proxy.c
@@ -138,7 +138,11 @@ test_basics (Fixture *f,
 
   alarm (30);
 
+#if GLIB_CHECK_VERSION (2, 78, 0)
+  g_unix_open_pipe (sync_pipe, O_CLOEXEC, &error);
+#else
   g_unix_open_pipe (sync_pipe, FD_CLOEXEC, &error);
+#endif
   g_assert_no_error (error);
   f->sync_pipe = sync_pipe[READ_END];
 
